### PR TITLE
Added podMonitor for kube-proxy

### LINKED ISSUE #321: Kubernetes / Proxy dashboard is empty
**What did you do?**
I've deployed prometheus operator, prometheus, grafana and alertmanager using kube-prometheus. I noticed that the "Kubernetes / Proxy" dashboard seemed to be empty

image

**Did you expect to see some different?**
I didn't expect an empty dashboard. The metric for the number of instances up shown in the graph was `sum(up{job="kube-proxy"})`. Upon debugging I saw that there was no ServiceMonitor in the generated manifests that had jobLabel set to something that had the value of kube-proxy.
I also noticed that the dropdown in the prometheus ui didn't have any metric with the prefix kubeproxy

Is this expected or is this something that is broken?

**Environment**

* Prometheus Operator version:

    `v0.33.0`
    

* Kubernetes version information:

    `Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:40:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:32:14Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}`
    

* Kubernetes cluster kind:

    kubeadm

**Anything else we need to know?**:
I noticed that the kube-proxy metrics are available on port 10249 with the metric keys present in this dashboard(for eg., kubeproxy_sync_proxy_rules_duration_seconds_count). But they are available only from localhost and not remotely. kube-proxy runs with hostNetwork set to true because of which I was able to ssh into the node and make a curl call on localhost and port 10249.

### PR DESCRIPTION
**Pull request by lyveng**
**Why have I added the kube-proxy pod monitor inside jsonnet/kube-prometheus/prometheus/prometheus.libsonnet?**
CoreDNS and kube-proxy are addons that are installed by default when kubeadm is used to bootstrap a cluster. I've added pod monitor for kube-proxy in prometheus.jsonnet since the service monitor object for CoreDNS was present in prometheus.jsonnet. AFAIK since kube-proxy is installed by default, it would be ok for the pod monitor also to be present by default when someone uses kube-prometheus.

**Couple of caveats that I wanted to discuss before I added documentation:**
1. Metrics bind address needs to be set to 0.0.0.0:10249. kube-proxy configmap can be updated and the kube-proxy daemonset can be restarted for existing clusters. [KubeProxyConfiguration](https://pkg.go.dev/k8s.io/kubernetes/pkg/proxy/apis/config?tab=doc#KubeProxyConfiguration) can be used in kubeadm config file to set the value of MetricsBindAddress to 0.0.0.0:10249
2. ~~Another problem is that, I've set targetPort in the pod monitor spec. But targetPort is deprecated as per the [prometheus operator docs](https://github.com/coreos/prometheus-operator/blob/327f2666938eddfa5ab0adb15119edb8e1664d63/Documentation/api.md#podmetricsendpoint) and port field is advised to be used. But we can't use port since the [kube-proxy daemonset manifest](https://github.com/kubernetes/kubernetes/blob/cab89c7d08a5a9ae4091a1aaaa9b9c0ab0f307da/cmd/kubeadm/app/phases/addons/proxy/manifests.go#L54) doesn't have a named port in the pod spec. If we still need to use port instead of targetPort, then we might have to add it as a step in documentation to edit the kube-proxy daemonset and add the named port manually.~~
**Update**: Setting targetPort in `PodMonitor` doesn't fix the service discovery. Port has to be added to the kube-proxy manifest for service discovery to work. Refer to this [comment](#issuecomment-622584604).

Fixes #321 
There is already another PR #464 that does the same task via service monitor.

### CODE REVIEW

**Comment by lyveng** [comment]
@brancz FYI

**Comment by lyveng** [comment]
I've tested this on cluster setup using kubeadm. I'm yet to test it on minikube. I'll update here once I've tested on minikube.

**Comment by lyveng** [comment]
@brancz While testing on minikube, I realized that setting targetPort also doesn't help with service discovery. This had worked earlier for me in my cluster bootstrapped via kubeadm because I had modified the kube-proxy daemonset to include the port definition when I was testing out port in podMonitor.spec.podMetricsEndpoint and had forgotten to remove it when I switched to targetPort. Since targetPort was missing in minikube, the service discovery for kube-proxy didn't match any pods.

If we want to use `PodMonitor`, then we'll have to either inform users to update their kube-proxy daemonset manifest to include a port for metrics or update the [upstream code](https://github.com/kubernetes/kubernetes/blob/cab89c7d08a5a9ae4091a1aaaa9b9c0ab0f307da/cmd/kubeadm/app/phases/addons/proxy/manifests.go#L54) to include the metrics port as part of the documentation.

This would be the same case even if we want to move controller-manager or scheduler to `PodMonitor` from `ServiceMonitor` as you had suggested in [#464](https://github.com/coreos/kube-prometheus/pull/464#discussion_r396918803).

**Comment by brancz** [comment]
Adding the metrics port to the `Pod` upstream seems reasonable to me.

File: jsonnet/kube-prometheus/prometheus/prometheus.libsonnet [hunk 1]
```diff
@@ -462,5 +462,36 @@ local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
...
+    podMonitorKubeProxy:
+      {
+        apiVersion: "monitoring.coreos.com/v1",
+        kind: "PodMonitor",
+        metadata: {
+          labels: {
+            "k8s-app": "kube-proxy"
+          },
+          name: "kube-proxy",
+          namespace: "monitoring"
```
Comment (errm) [inline jsonnet/kube-prometheus/prometheus/prometheus.libsonnet:474]:
```suggestion
          namespace: $._config.namespace
```

**Comment by cofyc** [comment]
@lyveng what's the progress?

**Comment by lyveng** [comment]
@cofyc I haven't gotten time to get this done.

**Comment by paulfantom** [comment]
Done with #1230

**Pull request closed without merging**